File: risk-engine/src/kinetix_risk/pnl_attribution_metrics.py
"""Prometheus instrumentation for Greek-component P&L attribution.

Records a completed :class:`~kinetix_risk.pnl_attribution.GreekPnlAttribution`
into the ``pnl_attribution_greek_pnl`` (per-Greek), ``pnl_attribution_dollar_delta``
and ``pnl_attribution_dollar_gamma`` gauges so Grafana can show a desk whether a
move was delta-, gamma-, vega- or theta-driven.
"""

from kinetix_risk.metrics import (
    pnl_attribution_dollar_delta,
    pnl_attribution_dollar_gamma,
    pnl_attribution_greek_pnl,
)
from kinetix_risk.pnl_attribution import GreekPnlAttribution


def record_greek_pnl_attribution(result: GreekPnlAttribution, book_id: str) -> None:
    """Set every P&L-decomposition gauge from a completed attribution result.

    Populates one ``pnl_attribution_greek_pnl`` series per Greek component plus
    the ``pnl_attribution_dollar_delta`` / ``pnl_attribution_dollar_gamma``
    cash-sensitivity gauges for the given book.
    """
    pnl_attribution_greek_pnl.labels(book_id=book_id, greek="delta").set(result.delta_pnl)
    pnl_attribution_greek_pnl.labels(book_id=book_id, greek="gamma").set(result.gamma_pnl)
    pnl_attribution_greek_pnl.labels(book_id=book_id, greek="vega").set(result.vega_pnl)
    pnl_attribution_greek_pnl.labels(book_id=book_id, greek="theta").set(result.theta_pnl)
    pnl_attribution_greek_pnl.labels(book_id=book_id, greek="rho").set(result.rho_pnl)

    pnl_attribution_dollar_delta.labels(book_id=book_id).set(result.dollar_delta)
    pnl_attribution_dollar_gamma.labels(book_id=book_id).set(result.dollar_gamma)

File: risk-orchestrator/src/main/kotlin/com/kinetix/risk/orchestrator/metrics/PnlAttributionMetrics.kt
package com.kinetix.risk.orchestrator.metrics

import com.kinetix.risk.model.PnlAttribution
import io.micrometer.core.instrument.Gauge
import io.micrometer.core.instrument.MeterRegistry
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.atomic.AtomicReference

/**
 * Publishes a book's daily P&L attribution as Prometheus gauges so the Grafana
 * "P&L" dashboard resolves (kx-wrvc).
 *
 * The dashboard queries `pnl_attribution_total_pnl`, `pnl_attribution_unexplained_pnl`
 * and `pnl_attribution_greek_pnl{greek=...}`. Those gauges were only ever set
 * inside the risk-engine's legacy `CalculateVaR` gRPC handler (superseded by the
 * unified `Valuate` RPC, ADR-0024) and its factor server — neither of which the
 * running platform drives — so the series never existed and every P&L panel read
 * "No data". The attribution itself is already computed and persisted by the
 * orchestrator (it backs the P&L tab), so here we simply mirror the persisted
 * numbers onto gauges the registry already scrapes.
 *
 * Each (book, greek) series is backed by an [AtomicReference] holder registered
 * once and updated in place, so repeated [publish] calls re-point the existing
 * gauge rather than registering duplicates.
 */
class PnlAttributionMetrics(private val registry: MeterRegistry) {

    private val totalPnl = ConcurrentHashMap<String, AtomicReference<Double>>()
    private val unexplainedPnl = ConcurrentHashMap<String, AtomicReference<Double>>()
    private val greekPnl = ConcurrentHashMap<String, AtomicReference<Double>>()

    fun publish(attribution: PnlAttribution) {
        val book = attribution.bookId.value
        holder(totalPnl, "pnl_attribution_total_pnl", book).set(attribution.totalPnl.toDouble())
        holder(unexplainedPnl, "pnl_attribution_unexplained_pnl", book).set(attribution.unexplainedPnl.toDouble())

        greekHolder(book, "delta").set(attribution.deltaPnl.toDouble())
        greekHolder(book, "gamma").set(attribution.gammaPnl.toDouble())
        greekHolder(book, "vega").set(attribution.vegaPnl.toDouble())
        greekHolder(book, "theta").set(attribution.thetaPnl.toDouble())
        greekHolder(book, "rho").set(attribution.rhoPnl.toDouble())
    }

    private fun holder(
        map: ConcurrentHashMap<String, AtomicReference<Double>>,
        metricName: String,
        book: String,
    ): AtomicReference<Double> =
        map.computeIfAbsent(book) {
            val ref = AtomicReference(0.0)
            Gauge.builder(metricName, ref) { it.get() }
                .tag("book_id", book)
                .register(registry)
            ref
        }

    private fun greekHolder(book: String, greek: String): AtomicReference<Double> =
        greekPnl.computeIfAbsent("$book|$greek") {
            val ref = AtomicReference(0.0)
            Gauge.builder("pnl_attribution_greek_pnl", ref) { it.get() }
                .tag("book_id", book)
                .tag("greek", greek)
                .register(registry)
            ref
        }
}
